### ISSUE TITLE
"Can not resolve reCAPTCHA parameter" entries in exception.log for Google reCAPTCHA Admin Panel

### ISSUE THREAD

**Issue by stevekem** [magento/magento2#34975]
### Preconditions (*)
1. Magento 2.4.3-p1

### Steps to reproduce (*)
1. Go to Configuration > Security > Google reCAPTCHA Admin Panel

2. Add your Google API Website Key & Google API Secret Key to any reCAPTCHA below:

reCAPTCHA v2 ("I am not a robot")
reCAPTCHA v2 Invisible
reCAPTCHA v3 Invisible

3. Select your reCAPTCHA version (with API keys above) for either below:

Enable for Login
Enable for Forgot Password

4. Save config

5. Flush cache

### Expected result (*)
1. No entry in var/log/exception.log file

### Actual result (*)
1. Below entry every few seconds/minutes constantly with no traffic:

main.ERROR: Can not resolve reCAPTCHA parameter. {"exception":"[object] (Magento\\Framework\\Exception\\InputException(code: 0): Can not resolve reCAPTCHA parameter. at /home/xxxxxxx/public_html/vendor/magento/module-re-captcha-ui/Model/CaptchaResponseResolver.php:25)"} []

- Getting around 12,000 entry lines every 24 hours in exception.log file which is getting huge in a short period of time.
- Keys have been triple checked to make sure correct, even created new ones with no change in behavior.
-  reCAPTCHA for both Admin and Frontend seems to be working.
- Only known way to stop entries is disable both "Enable for Login" and "Enable for Forgot Password" for Admin Storefront.
- reCAPTCHA for Frontend storefront isn't experiencing this issue for us for all enabled areas.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by 0xMatt** [comment]
Same issue. Happens all the time.

```
[2022-01-17 14:31:45] main.ERROR: Can not resolve reCAPTCHA parameter. {"exception":"[object] (Magento\\Framework\\Exception\\InputException(code: 0): Can not resolve reCAPTCHA parameter. at /var/www/html/magento/vendor/magento/module-re-captcha-ui/Model/CaptchaResponseResolver.php:25)"} []

[2022-01-17 16:21:40] main.ERROR: Can not resolve reCAPTCHA parameter. {"exception":"[object] (Magento\\Framework\\Exception\\InputException(code: 0): Can not resolve reCAPTCHA parameter. at /var/www/html/magento/vendor/magento/module-re-captcha-ui/Model/CaptchaResponseResolver.php:25)"} []
```

Captcha works. The badge is loaded and visible. When I go to log in, I'll get an error message saying something went wrong. I'll have to log in 2-3 times and around the third time it'll let me in. No javascript console errors are present during this issue.

**Comment by stevekem** [comment]
> Captcha works. The badge is loaded and visible. When I go to log in, I'll get an error message saying something went wrong. I'll have to log in 2-3 times and around the third time it'll let me in. No javascript console errors are present during this issue.

Yes this happens as well,  I forgot to mention that. Thank you!

**Comment by stevekem** [comment]
Just wanted to note that we are getting the "Something went wrong with reCAPTCHA. Please contact the store owner." error on frontend too, not just on Admin.  The 12,000+ daily log entries are happening with Admin module though.

**Comment by ihor-sviziev** [comment]
We got the same issue, but after analyzing it - it looks like we're getting a lot of such messages because some bot(s) are trying to create a customer, and captcha validation fails for them. We weren't able to reproduce this issue ourselves.

**Comment by stevekem** [comment]
> We got the same issue, but after analyzing it - it looks like we're getting a lot of such messages because some bot(s) are trying to create a customer, and captcha validation fails for them. We weren't able to reproduce this issue ourselves.

Thanks for the info! How could we verify if that is the case for us as well?

**Comment by ihor-sviziev** [comment]
@stevekem, we analyzed logs, the interval between requests from the same IP address was ~0-3 sec (usually people don't do requests so quickly), the requests were to `/customer/account/create/`, `/customer/account/createpost/` and to `/customer/account/create/<product_url_path>/`.
The last case links are not available on the website, so that's some bot for sure.

**Comment by stevekem** [comment]
@ihor-sviziev Ok gotcha!  If it is the bots, just block them in robots? Did that resolve the issue for you?

**Comment by ihor-sviziev** [comment]
@stevekem that looks like boys that are trying to hack or do something similar by creating a new customer

**Comment by stevekem** [comment]
@ihor-sviziev In your case, did you block the bots in robots.txt to resolve the issue?

**Comment by ihor-sviziev** [comment]
@stevekem yes, the requests starting with `/customer/` are blocked by default and we're not removed that.

**Comment by stevekem** [comment]
@ihor-sviziev Thank you, appreciate it! Will check our logs and see if that's the case for us as well.

**Comment by engcom-Delta** [comment]
Hi @stevekem ,
Thank you for reporting the issue,We are unable to  reproduce the issue on Magento 2.4 develop branch.Please review provided steps.
Steps Followed:

1.Login to admin Panel
2.Go to Stores-->Configuration-->Security-->Google Recaptcha Admin Panel-->reCAPTCHA v2("I am not a robot)
3.Added Google API Website Key and Google API Secret Key to reCAPTCHA v2("I am not a robot)
4.Enabled for Login
Enabled for Forgot Password
5.Saved the configuration
6.Flush Cache

**Actual Results** 
1.No entry in var/log/exception.log file
![Screenshot from 2022-02-23 20-11-06](https://user-images.githubusercontent.com/51681379/156531039-839732b8-a1f9-4cc1-b4b8-bb189937a6d5.png)

Hence Adding the label "Needs Update"

**Comment by engcom-Delta** [comment]
We have noticed that this issue has not been updated for a period of more than 14 Days. 
Hence we assume that this issue is fixed now, so we are closing it. Please raise a fresh ticket or reopen this ticket if you need more assistance on this.

**Comment by Hanhan1989** [comment]
Same problem in Magento 2.4.4.
`
 main.ERROR: Magento\Framework\Exception\InputException: Can not resolve reCAPTCHA parameter. in vendor/magento/module-re-captcha-ui/Model/CaptchaResponseResolver.php:25`

**Comment by Hanhan1989** [comment]
In my case, there were 2 javascript errors that caused me to not be able to log in. These errors are due to the minification of the **collapse.js** and **tab.js** files in the **lib/web/jquery/bootstrap/** path.

So the solution to this problem has been to exclude those files.

```
<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Store:etc/config.xsd">
    <default>
        <dev>
            <js>
                <minify_exclude>
                    <collapse>/jquery/bootstrap/collapse</collapse>
                    <tab>/jquery/bootstrap/tab</tab>
                </minify_exclude>
            </js>
        </dev>
    </default>
</config>
```

**Comment by Themageguru** [comment]
Hello, We got the same issue. Any update on this issue.

**Comment by mikeastral** [comment]
@engcom-Delta @knoorani-ijona I am having the same problem in Magento 2.4.4. 
I'll have to log in 2-3 times before it allows me to log in.  This issue is not occurring on frontend side.

 Minification of the collapse.js and tab.js files in the lib/web/jquery/bootstrap/ path is not fixing the issue.

**Comment by Hanhan1989** [comment]
> @engcom-Delta @knoorani-ijona I am having the same problem in Magento 2.4.4. 
> 
> I'll have to log in 2-3 times before it allows me to log in.  This issue is not occurring on frontend side.
> 
> 
> 
>  Minification of the collapse.js and tab.js files in the lib/web/jquery/bootstrap/ path is not fixing the issue.

I have to disable recaptcha for admin login。I hope magento open this issue and review the problem。thanks

**Comment by benediktblank** [comment]
I'm having the same issue. On Friday I upgraded from Magento 2.4.3-p1 to 2.4.4 and today I'm unable to log in to the Backend/Admin. I had to disable Admin Login Recaptcha in the Database to be able to login again.

I have no idea what is going on and have to keep recaptcha disabled in the Admin login for now.

Stack trace is:
[2022-05-23T17:10:05.800157+00:00] main.ERROR: Magento\Framework\Exception\InputException: Can not resolve reCAPTCHA parameter. in /var/www/html/vendor/magento/module-re-captcha-ui/Model/CaptchaResponseResolver.php:25
Stack trace:
#0 /var/www/html/vendor/magento/module-re-captcha-user/Observer/LoginObserver.php(125): Magento\ReCaptchaUi\Model\CaptchaResponseResolver->resolve(Object(Magento\Framework\App\Request\Http))
#1 /var/www/html/vendor/magento/framework/Event/Invoker/InvokerDefault.php(88): Magento\ReCaptchaUser\Observer\LoginObserver->execute(Object(Magento\Framework\Event\Observer))
#2 /var/www/html/vendor/magento/framework/Event/Invoker/InvokerDefault.php(74): Magento\Framework\Event\Invoker\InvokerDefault->_callObserverMethod(Object(Magento\ReCaptchaUser\Observer\LoginObserver), Object(Magento\Framework\Event\Observer))
#3 /var/www/html/vendor/magento/framework/Event/Manager.php(66): Magento\Framework\Event\Invoker\InvokerDefault->dispatch(Array, Object(Magento\Framework\Event\Observer))
#4 /var/www/html/generated/code/Magento/Framework/Event/Manager/Proxy.php(95): Magento\Framework\Event\Manager->dispatch('admin_user_auth...', Array)
#5 /var/www/html/vendor/magento/module-user/Model/User.php(589): Magento\Framework\Event\Manager\Proxy->dispatch('admin_user_auth...', Array)
#6 /var/www/html/generated/code/Magento/User/Model/User/Interceptor.php(140): Magento\User\Model\User->authenticate('benedikt.blank', 'sX6!yK9uv3fu')
#7 /var/www/html/vendor/magento/module-user/Model/User.php(649): Magento\User\Model\User\Interceptor->authenticate('benedikt.blank', 'sX6!yK9uv3fu')
#8 /var/www/html/generated/code/Magento/User/Model/User/Interceptor.php(158): Magento\User\Model\User->login('benedikt.blank', 'sX6!yK9uv3fu')
#9 /var/www/html/vendor/magento/module-backend/Model/Auth.php(162): Magento\User\Model\User\Interceptor->login('benedikt.blank', 'sX6!yK9uv3fu')
#10 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Backend\Model\Auth->login('benedikt.blank', 'sX6!yK9uv3fu')
#11 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Backend\Model\Auth\Interceptor->___callParent('login', Array)
#12 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Backend\Model\Auth\Interceptor->Magento\Framework\Interception\{closure}('benedikt.blank', 'sX6!yK9uv3fu')
#13 /var/www/html/generated/code/Magento/Backend/Model/Auth/Interceptor.php(59): Magento\Backend\Model\Auth\Interceptor->___callPlugins('login', Array, Array)
#14 /var/www/html/vendor/magento/module-backend/App/Action/Plugin/Authentication.php(207): Magento\Backend\Model\Auth\Interceptor->login('benedikt.blank', 'sX6!yK9uv3fu')
#15 /var/www/html/vendor/magento/module-backend/App/Action/Plugin/Authentication.php(159): Magento\Backend\App\Action\Plugin\Authentication->_performLogin(Object(Magento\Framework\App\Request\Http))
#16 /var/www/html/vendor/magento/module-backend/App/Action/Plugin/Authentication.php(127): Magento\Backend\App\Action\Plugin\Authentication->_processNotLoggedInUser(Object(Magento\Framework\App\Request\Http))
#17 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(135): Magento\Backend\App\Action\Plugin\Authentication->aroundDispatch(Object(Magento\Backend\Controller\Adminhtml\Index\Index\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#18 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Backend\Controller\Adminhtml\Index\Index\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#19 /var/www/html/generated/code/Magento/Backend/Controller/Adminhtml/Index/Index/Interceptor.php(32): Magento\Backend\Controller\Adminhtml\Index\Index\Interceptor->___callPlugins('dispatch', Array, Array)
#20 /var/www/html/vendor/magento/framework/App/FrontController.php(245): Magento\Backend\Controller\Adminhtml\Index\Index\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#21 /var/www/html/vendor/magento/framework/App/FrontController.php(212): Magento\Framework\App\FrontController->getActionResponse(Object(Magento\Backend\Controller\Adminhtml\Index\Index\Interceptor), Object(Magento\Framework\App\Request\Http))
#22 /var/www/html/vendor/magento/framework/App/FrontController.php(147): Magento\Framework\App\FrontController->processRequest(Object(Magento\Framework\App\Request\Http), Object(Magento\Backend\Controller\Adminhtml\Index\Index\Interceptor))
#23 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Framework\App\FrontController->dispatch(Object(Magento\Framework\App\Request\Http))
#24 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\App\FrontController\Interceptor->___callParent('dispatch', Array)
#25 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#26 /var/www/html/generated/code/Magento/Framework/App/FrontController/Interceptor.php(23): Magento\Framework\App\FrontController\Interceptor->___callPlugins('dispatch', Array, Array)
#27 /var/www/html/vendor/magento/framework/App/Http.php(116): Magento\Framework\App\FrontController\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#28 /var/www/html/generated/code/Magento/Framework/App/Http/Interceptor.php(23): Magento\Framework\App\Http->launch()
#29 /var/www/html/vendor/magento/framework/App/Bootstrap.php(264): Magento\Framework\App\Http\Interceptor->launch()
#30 /var/www/html/pub/index.php(30): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http\Interceptor))
#31 {main} [] []

**Comment by tschirmer** [comment]
The reason for this is that the JavaScript client isn't getting the token on submission. So the ReCaptcha param it's looking for is empty. 

The client isn't resolving correctly, because it's being loaded asynchronously and the way it's coded, will only activate the form interception after the 
vendor/magento/module-re-captcha-user/view/adminhtml/templates/recaptcha.phtml


change       
```
        const element = document.createElement('script');
        const scriptTag = document.getElementsByTagName('script')[0];

        element.async = true;
        element.src = 'https://www.google.com/recaptcha/api.js'
            + '?onload=globalOnRecaptchaOnLoadCallback&render=explicit';

```
to
```
        const element = document.createElement('script');
        const scriptTag = document.getElementsByTagName('script')[0];

        element.async = false;
        element.src = 'https://www.google.com/recaptcha/api.js'
            + '?onload=globalOnRecaptchaOnLoadCallback&render=explicit';

```

Let me know if that patch works. Technically it'll load slower, but slow and working is always better than fast and broken.

**Comment by tschirmer** [comment]
This is the offending file.

vendor/magento/module-re-captcha-user/view/adminhtml/templates/recaptcha.phtml

Essentially, this was coded in a way that the submit function binders were able to allow the submit wihtout waiting for the recaptcha code. 

I've recoded this, so replace that file with this and it should work. This might need extra work to make sure the invisible captcha configuration works as expected.


```
<?php /** * Copyright © Magento, Inc. All rights reserved. * See COPYING.txt for license details. */ /** @var $block Magento\ReCaptchaUi\Block\ReCaptcha */ $config = $block->getCaptchaUiConfig(); $renderingOptions = $config['rendering'] ?? []; $isInvisible = !empty($config['invisible']); ?> <div class="admin__field <?= /* @noEscape */ $isInvisible ? 'field-invisible-recaptcha' : 'field-recaptcha' ?>"><div id="admin-recaptcha" class="admin-recaptcha-content<?= /* @noEscape */ !empty($renderingOptions['size']) ? ' size-' . $renderingOptions['size'] : '' ?>"></div></div><script>
    require([
        'jquery'
    ], function (
        $
    ) {
        const element = document.createElement('script');
        const scriptTag = document.getElementsByTagName('script')[0];

        element.async = true;
        element.src = 'https://www.google.com/recaptcha/api.js'
            + '?onload=globalOnRecaptchaOnLoadCallback&render=explicit';

        let isRecaptchaLoaded = false;
        let token = '';
        let maxRetryAttempts = 5;
        let attempts = 0;

        $('#login-form').submit(function (event) {
            if (!token) {
                event.preventDefault(event);
                event.stopPropagation();

                let attemptRecaptcha = () => {
                    attempts++;
                    if (attempts > maxRetryAttempts){
                        console.error("Could not fetch invisible ReCaptcha token. Please refresh the page and try again.");
                        return;
                    }
                    if (!isRecaptchaLoaded) {
                        //wait until it is
                        setTimeout(() => {
                            attemptRecaptcha()
                        }, 1000);
                        return;
                    }
                    let widgetId = grecaptcha.render('admin-recaptcha', {
                        <?php foreach ($renderingOptions as $key => $value): ?> '<?= $block->escapeJs($key) ?>': '<?= $block->escapeJs($value) ?>',
                        <?php endforeach; ?> 'callback': function (_token) {
                            <?php if ($isInvisible): ?>
                            token = _token;
                            $('#login-form').unbind('submit');
                            $('#login-form').submit();
                            <?php endif; ?> }
                    });
                    grecaptcha.execute(widgetId)
                        .then( () => {
                            event.preventDefault(event);
                            event.stopImmediatePropagation();
                            event.stopPropagation();
                        }, (reason) => { })
                        .catch(err => { console.error(err); });
                }
                attemptRecaptcha();
            }
        });
        window.globalOnRecaptchaOnLoadCallback = function () {
            isRecaptchaLoaded = true;
        }
        scriptTag.parentNode.insertBefore(element, scriptTag);
    });</script>
```

**Comment by C4rter** [comment]
Wow, thank you. Yeah, it works with that code now indeed.

The blue "protected by recaptcha" box now only appears briefly after submitting the form. It's not visible before.
Is that part of the fix?

The "g-recaptcha-response" parameter is now correctly filled again and the exception is therefore not thrown anymore.

Just noticed I accidentally made the other post with an old account of mine.

**Comment by mikeastral** [comment]
Excellent,

This last code completely fixed the re-captcha issue

Thank you very much

**Comment by C4rter** [comment]
@tschirmer 

Thank you again! That does it. The Recaptcha Box is visible and the login works.

I've created a composer patch with your changes, if anybody else needs to apply that fix during deployment with cweagans/composer-patches or something else:
```
diff --git a/view/adminhtml/templates/recaptcha.phtml b/view/adminhtml/templates/recaptcha.phtml
index 1f423e801087..0e4d79aac933 100644
--- a/view/adminhtml/templates/recaptcha.phtml
+++ b/view/adminhtml/templates/recaptcha.phtml
@@ -27,36 +27,56 @@
         element.src = 'https://www.google.com/recaptcha/api.js'
             + '?onload=globalOnRecaptchaOnLoadCallback&render=explicit';

-        window.globalOnRecaptchaOnLoadCallback = function () {
-            let token = '';
+        let isRecaptchaLoaded = false;
+        let token = '';
+        let maxRetryAttempts = 5;
+        let attempts = 0;
+        let widgetId = 0;
+        <?php if ($isInvisible): ?>
+        $('#login-form').submit(function (event) {
+            if (!token) {
+                event.preventDefault(event);
+                event.stopImmediatePropagation();
+                event.stopPropagation();

-            this.widgetId = grecaptcha.render('admin-recaptcha', {
-                <?php foreach ($renderingOptions as $key => $value): ?>
-                '<?= $block->escapeJs($key) ?>': '<?= $block->escapeJs($value) ?>',
-                <?php endforeach; ?>
-                'callback': function (token) { // jscs:ignore jsDoc
-                    <?php if ($isInvisible): ?>
-                    this.token = token;
-                    $('#login-form').submit();
-                    <?php endif; ?>
-                }.bind(this)
-            });
+                let attemptRecaptcha = () => {
+                    attempts++;
+                    if (attempts > maxRetryAttempts){
+                        console.error("Could not fetch invisible ReCaptcha token. Please refresh the page and try again.");
+                        return;
+                    }
+                    if (!isRecaptchaLoaded) {

-            <?php if ($isInvisible): ?>
-            $('#login-form').submit(function (event) {
-                if (!this.token) {
-                    grecaptcha.execute(this.widgetId).then(
-                        function() {
+                        setTimeout(() => {
+                            attemptRecaptcha()
+                        }, 1000);
+                        return;
+                    }
+                    grecaptcha.execute(widgetId)
+                        .then( () => {
                             event.preventDefault(event);
                             event.stopImmediatePropagation();
-                        }, function(reason) {
-                        });
+                            event.stopPropagation();
+                        }, (reason) => { })
+                        .catch(err => { console.error(err); });
                 }
-            }.bind(this));
-            <?php endif; ?>
-
-        }.bind(this);
+                attemptRecaptcha();
+            }
+        });
+        <?php endif; ?>
+        window.globalOnRecaptchaOnLoadCallback = function () {

+            widgetId = grecaptcha.render('admin-recaptcha', {
+                <?php foreach ($renderingOptions as $key => $value): ?> '<?= $block->escapeJs($key) ?>': '<?= $block->escapeJs($value) ?>',
+                <?php endforeach; ?> 'callback': function (_token) {
+                    <?php if ($isInvisible): ?>
+                    token = _token;
+                    $('#login-form').unbind('submit');
+                    $('#login-form').submit();
+                    <?php endif; ?> }
+            });
+            isRecaptchaLoaded = true;
+        }
         scriptTag.parentNode.insertBefore(element, scriptTag);
     });
 </script>

```

**Comment by faisal-project** [comment]
Issue is also happening on frontend  (Magento CE 2.4.4)

main.ERROR: Magento\Framework\Exception\InputException: Can not resolve reCAPTCHA parameter. in /vendor/magento/module-re-captcha-ui/Model/CaptchaResponseResolver.php:25

Is there any solution for frontend?

Thanks!

**Comment by heyepic** [comment]
I also updated a site from 2.4.3-p1 to 2.4.4 and now the admin doesn't display captcha and I cannot get into the admin panel.  Nothing in my logs to go off of either. Super frustrated as I've spent too much time trying to figure out why. Didn't have issues like this prior - any ideas as to what's going on?

**Comment by knospe-** [comment]
I can confirm that this issue also occurs in Frontend in version 2.4.3-p12.

**Comment by drinkingsouls** [comment]
Just upgraded from 2.4.3-p3 to 2.4.4 and now experiencing this issue on Admin login. Can't login with Google Recaptcha

[2022-07-14T10:12:07.661109+00:00] main.ERROR: Magento\Framework\Exception\InputException: Can not resolve reCAPTCHA parameter. in /vendor/magento/module-re-captcha-ui/Model/CaptchaResponseResolver.php:25

**Comment by benediktblank** [comment]
> Just upgraded from 2.4.3-p3 to 2.4.4 and now experiencing this issue on Admin login. Can't login with Google Recaptcha
> 
> [2022-07-14T10:12:07.661109+00:00] main.ERROR: Magento\Framework\Exception\InputException: Can not resolve reCAPTCHA parameter. in /vendor/magento/module-re-captcha-ui/Model/CaptchaResponseResolver.php:25

You should apply the fix from @tschirmer or my composer patch that contains his fix.

**Comment by drinkingsouls** [comment]
The patch here fails to apply for me:
Corrupt patch at line 8.

**Comment by drinkingsouls** [comment]
Manually applied the code by @tschirmer
When using this, the Recaptcha box doesn't show until the Username and Password have been entered and Sign In has been clicked.
The Recaptcha box then momentarily shows before proceeding to the 2FA Authenticator code step.

It feels a little like its bypassing Google Recaptcha completely with the way it's shown. Can anyone confirm that the Recaptcha works correctly with this patch?

**Comment by benediktblank** [comment]
I think you applied his first version. You should apply the "Try this instead.".
I mentioned the same behavior already you describe and he provided an adjusted fix.

**Comment by cjohansson** [comment]
Same issue here with Magento 2.4.4, the issue seems to code and go

**Comment by cjohansson** [comment]
It seems to have something to do with reputation or session/cookie, if I change my IP and clear all cookies I am able to login. There might be a error just in the message of the error

**Comment by engcom-Hotel** [comment]
Hello,

As I can see this issue got fixed in the scope of the internal Jira ticket AC-3179 by the internal team
Related commits: https://github.com/search?q=AC-3179&type=commits

Based on the Jira ticket, the target version is 2.4.6.

Thanks

**Comment by mikeastral** [comment]
Changing the file mentioned is still not fixing the issue.
After directly changing the file in the vendor folder, just for
testing purposes, I compile and deploy static-static content, but the issue
is occurring at the same rate.

bin/magento setup:upgrade;
bin/magento setup:di:compile;
bin/magento setup:static-content:deploy;
bin/magento cache:flush;

**Comment by mrramanathan31** [comment]
@stevekem Getting the same below error on Magento 2.4.5-p1 Enterprise edition. Please provide the patch for this issue.

`report.ERROR: Magento\Framework\Exception\InputException: Can not resolve reCAPTCHA parameter. in /var/www/html/oreo/vendor/magento/module-re-captcha-ui/Model/CaptchaResponseResolver.php:25`

**Comment by briangmoore77** [comment]
Having to wait 4 months (3/14/23) for this fix to be in 2.4.6 is a little ridiculous.

**Comment by Hanhan1989** [comment]
sadly, my solution was disable it.

**Comment by tschirmer** [comment]
yeah it's a long time between writing the bugfix and getting into a release version. Would be good to get attribution on my work there @engcom-Hotel

**Comment by briangmoore77** [comment]
@Hanhan1989 you disabled the recaptcha completely?

**Comment by Hanhan1989** [comment]
> @Hanhan1989 you disabled the recaptcha completely?

I only disabled the Recaptcha of the admin user login form. 

`bin/magento security:recaptcha:disable-for-user-login `

**Comment by briangmoore77** [comment]
@Hanhan1989 you don't have bots hitting your other login pages causing the same error?

**Comment by Hanhan1989** [comment]
@briangmoore77  I only had problem of Recaptcha in the admin user login form. In the Frontend， all Ok.

**Comment by hprasetyou** [comment]
Weird for me, the error message are there on the system.log, but i can still logged in on chrome and safari, but cannot be logged in on firefox

**Comment by rossparachute** [comment]
v2.4.5-p1 - bug still present on the frontend forms.

Same exception:
`Magento\Framework\Exception\InputException: Can not resolve reCAPTCHA parameter. in /vendor/magento/module-re-captcha-ui/Model/CaptchaResponseResolver.php:25
`

The actual ReCaptcha itself renders but I'm not seeing a recaptcha-response field being added to the form that could then be used to validate the form submission.

Notable forms:
- Login
- Forgotten Password
- Register

@tschirmer fix - totally welcome but doesn't fix the frontend issue (though, wasn't intended to as far as I can see). 

Oddly, the captcha is on the Checkout form too and that appears to be behaving fine.

**Comment by peterjaap** [comment]
@engcom-Alfa @engcom-Bravo @engcom-Charlie @engcom-Delta @engcom-Hotel it seems this isn't fixed at all. 

Maybe in 2.4.6 but then we'd like to see a commit with the fix so we can patch our stores. Please advise.

**Comment by rossparachute** [comment]
@peterjaap 

Your mileage may vary but we found we had to manually add in the ReCaptcha block back into the forms.

I think there's some move/remove statement somewhere in Magento or a module we all have in common that is taking out the block named "recaptcha" that is intended to be output in the _form_additional_info_ child block.

So we had to create the following under _<theme_name>/Magento_ReCaptchaCustomerLayout_:

1. customer_account_create.xml
2. customer_account_edit.xml
3. customer_account_forgotpassword.xml
4. customer_account_login.xml
5. default.xml

The form layout files were all pretty much the same:

```
<?xml version="1.0"?>
<!--
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */
-->
<page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
    <body>
        <!-- Required to ensure the ReCaptcha actually gets added - something is taking it out in 2.4.5 -->
        <!-- Note renaming the block from "recaptcha" to "recaptcha.magento" -->
        <referenceContainer name="customer_form_edit">
            <block class="Magento\ReCaptchaUi\Block\ReCaptcha"
                   name="recaptcha.magento"
                   after="-"
                   template="Magento_ReCaptchaFrontendUi::recaptcha.phtml"
                   ifconfig="recaptcha_frontend/type_for/customer_edit">
                <arguments>
                    <argument name="recaptcha_for" xsi:type="string">customer_edit</argument>
                    <argument name="jsLayout" xsi:type="array">
                        <item name="components" xsi:type="array">
                            <item name="recaptcha" xsi:type="array">
                                <item name="component" xsi:type="string">Magento_ReCaptchaFrontendUi/js/reCaptcha</item>
                            </item>
                        </item>
                    </argument>
                </arguments>
            </block>
        </referenceContainer>
    </body>
</page>
```

Then in your form, output the child HTML block:

```
<form>
...
<?= $block->getChildHtml('recaptcha.magento') ?>
...
</form>
```

P.S For what it's worth we're also using the _mageplaza/module-google-recaptcha_ package (v^4.0) for the admin which can do ReCaptcha too although that's disabled on the frontend. When I looked at the module's layout XML files I couldn't see it moving/removing anything. I don't think that's affecting things but it's worth nothing.

**Comment by engcom-Hotel** [comment]
@peterjaap I have provided the related commits for this fix here https://github.com/magento/magento2/issues/34975#issuecomment-1246619119. Does it help in moving further with this?

**Comment by BernardRobbins-BMediaLLC** [comment]
I'm also getting this error flooding my logs:

```
Magento\Framework\Exception\InputException: Can not resolve reCAPTCHA parameter. in /var/www/html/vendor/magento/module-re-captcha-ui/Model/CaptchaResponseResolver.php:25
Stack trace:
#0 /var/www/html/vendor/magento/module-re-captcha-ui/Model/RequestHandler.php(112): Magento\ReCaptchaUi\Model\CaptchaResponseResolver->resolve()
#1 /var/www/html/vendor/magento/module-re-captcha-customer/Observer/CreateCustomerObserver.php(67): Magento\ReCaptchaUi\Model\RequestHandler->execute()
#2 /var/www/html/vendor/magento/framework/Event/Invoker/InvokerDefault.php(88): Magento\ReCaptchaCustomer\Observer\CreateCustomerObserver->execute()
#3 /var/www/html/vendor/magento/framework/Event/Invoker/InvokerDefault.php(74): Magento\Framework\Event\Invoker\InvokerDefault->_callObserverMethod()
#4 /var/www/html/vendor/magento/framework/Event/Manager.php(65): Magento\Framework\Event\Invoker\InvokerDefault->dispatch()
#5 /var/www/html/generated/code/Magento/Framework/Event/Manager/Proxy.php(95): Magento\Framework\Event\Manager->dispatch()
#6 /var/www/html/vendor/magento/framework/App/FrontController.php(279): Magento\Framework\Event\Manager\Proxy->dispatch()
#7 /var/www/html/vendor/magento/framework/App/FrontController.php(211): Magento\Framework\App\FrontController->dispatchPreDispatchEvents()
#8 /var/www/html/vendor/magento/framework/App/FrontController.php(147): Magento\Framework\App\FrontController->processRequest()
#9 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Framework\App\FrontController->dispatch()
#10 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\App\FrontController\Interceptor->___callParent()
#11 /var/www/html/vendor/magento/module-store/App/FrontController/Plugin/RequestPreprocessor.php(99): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}()
#12 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(135): Magento\Store\App\FrontController\Plugin\RequestPreprocessor->aroundDispatch()
#13 /var/www/html/vendor/magento/module-page-cache/Model/App/FrontController/BuiltinPlugin.php(75): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}()
#14 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(135): Magento\PageCache\Model\App\FrontController\BuiltinPlugin->aroundDispatch()
#15 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}()
#16 /var/www/html/generated/code/Magento/Framework/App/FrontController/Interceptor.php(23): Magento\Framework\App\FrontController\Interceptor->___callPlugins()
#17 /var/www/html/vendor/magento/framework/App/Http.php(116): Magento\Framework\App\FrontController\Interceptor->dispatch()
#18 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Framework\App\Http->launch()
#19 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\App\Http\Interceptor->___callParent()
#20 /var/www/html/vendor/justbetter/magento2-sentry/Plugin/GlobalExceptionCatcher.php(78): Magento\Framework\App\Http\Interceptor->Magento\Framework\Interception\{closure}()
#21 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(135): JustBetter\Sentry\Plugin\GlobalExceptionCatcher->aroundLaunch()
#22 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Framework\App\Http\Interceptor->Magento\Framework\Interception\{closure}()
#23 /var/www/html/generated/code/Magento/Framework/App/Http/Interceptor.php(23): Magento\Framework\App\Http\Interceptor->___callPlugins()
#24 /var/www/html/vendor/magento/framework/App/Bootstrap.php(264): Magento\Framework\App\Http\Interceptor->launch()
#25 /var/www/html/pub/index.php(30): Magento\Framework\App\Bootstrap->run()
#26 {main}

```

Here's the nginx logs from the flood of bots:

```
103.233.254.62 - - [15/Feb/2023:16:19:56 +0000] "GET /customer/account/create HTTP/1.1" 200 41450 "https://www.{our_website}.com/customer/account/create" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.108 Safari/537.36"
103.233.254.62 - - [15/Feb/2023:16:19:56 +0000] "POST /customer/account/createpost/ HTTP/1.1" 302 5 "https://www.{our_website}.com/customer/account/create" "Mozilla/5.0 (Windows NT 6.3; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36"

```

@engcom-Hotel I've noticed your commits are for the admin login. Will this also somehow fix the frontend customer account creation?

**Comment by hostep** [comment]
@BernardRobbins-BMediaLLC, another [batch](https://github.com/magento/security-package/commit/deae52075d31ffea0a0235d13a02ba587021b969) of [commits](https://github.com/magento/magento2/commit/3218b12a60e6174dfe02aef28ba6532dc63988dc) got merged yesterday, that have something to do with recaptcha, maybe those help in your case? Just guessing here ...

**Comment by BernardRobbins-BMediaLLC** [comment]
@hostep, thanks, it doesn't appear any of those will resolve this.

**Comment by sebfie** [comment]
I still have this issue in 2.4.6

**Comment by quyendam2612** [comment]
I can confirm the issue still happens in 2.4.6

**Comment by OvalMedia** [comment]
I think as long as this thread is closed nobody will take notice.

**Comment by Hanhan1989** [comment]
I have Magento 2.4.6 and it works correctly。I think the error may be due to third party extensions。

**Comment by AlexandreBulete** [comment]
Hi, 

I didn't faced this issue in admin panel but in a front page. 
I saw that was caused by moving the **form.additional.info** container. This container contains a few blocks, including the recaptcha block.

You should verify if this block is present in your admin form by using a debugtoolbar as [Smile_DebugToolbar](https://github.com/Smile-SA/magento2-module-debug-toolbar)

My account login page include the form login and the form register then the container was moved in the register form and wasn't present in the login form.

Then I had to add this recaptcha block myself inside a custom **form.additional.info.login** container which is itself inside my login form : 
```xml
<container name="form.additional.info.login" as="form_additional_info_login">
    <block class="Magento\ReCaptchaUi\Block\ReCaptcha"
        name="recaptcha.login"
        after="-"
        template="Magento_ReCaptchaFrontendUi::recaptcha.phtml"
        ifconfig="recaptcha_frontend/type_for/customer_login"
    >
        <arguments>
            <argument name="recaptcha_for" xsi:type="string">customer_login</argument>
            <argument name="jsLayout" xsi:type="array">
                <item name="components" xsi:type="array">
                    <item name="recaptcha" xsi:type="array">
                        <item name="component" xsi:type="string">Magento_ReCaptchaFrontendUi/js/reCaptcha</item>
                    </item>
                </item>
            </argument>
        </arguments>
    </block>
</container>
```

**Comment by 0xMatt** [comment]
Still having this issue in 2.4.6

**Comment by perryholden** [comment]
The patch supplied above works for me on version 2.4.4-p5. Thanks to @tschirmer and @C4rter.

**Comment by frostitution** [comment]
Discovered this issue not too long after enabling official Magento reCAPTCHA Checkbox on Magento 2.4.6, still an issue

**Comment by AlexandreBulete** [comment]
Each form "Recaptchable" has an additional script block added. You have to watch the corresponding recaptcha module.
For example : module-re-captcha-newsletter add this block : 

```xml
<referenceContainer name="footer">
            <block class="Magento\ReCaptchaUi\Block\ReCaptcha"
                   name="recaptcha-newsletter"
                   after="-"
                   template="Magento_ReCaptchaNewsletter::recaptcha_newsletter.phtml"
                   ifconfig="recaptcha_frontend/type_for/newsletter">
                <arguments>
                    <argument name="recaptcha_for" xsi:type="string">newsletter</argument>
                    <argument name="jsLayout" xsi:type="array">
                        <item name="components" xsi:type="array">
                            <item name="recaptcha" xsi:type="array">
                                <item name="component" xsi:type="string">Magento_ReCaptchaFrontendUi/js/reCaptcha</item>
                            </item>
                        </item>
                    </argument>
                </arguments>
            </block>
        </referenceContainer>
```

Are you sure your layout is including this additional script block ?

I noticed it after dd() the `$request->getParams()` in this `CaptchaResponseResolver`model.
If you haven't the requested recaptcha params then it seems you didn't include this script block.

**Comment by rossparachute** [comment]
> Each form "Recaptchable" has an additional script block added. You have to watch the corresponding recaptcha module. For example : module-re-captcha-newsletter add this block :
> 
> ```
> <referenceContainer name="footer">
>             <block class="Magento\ReCaptchaUi\Block\ReCaptcha"
>                    name="recaptcha-newsletter"
>                    after="-"
>                    template="Magento_ReCaptchaNewsletter::recaptcha_newsletter.phtml"
>                    ifconfig="recaptcha_frontend/type_for/newsletter">
>                 <arguments>
>                     <argument name="recaptcha_for" xsi:type="string">newsletter</argument>
>                     <argument name="jsLayout" xsi:type="array">
>                         <item name="components" xsi:type="array">
>                             <item name="recaptcha" xsi:type="array">
>                                 <item name="component" xsi:type="string">Magento_ReCaptchaFrontendUi/js/reCaptcha</item>
>                             </item>
>                         </item>
>                     </argument>
>                 </arguments>
>             </block>
>         </referenceContainer>
> ```
> 
> Are you sure your layout is including this additional script block ?
> 
> I noticed it after dd() the `$request->getParams()` in this `CaptchaResponseResolver`model. If you haven't the requested recaptcha params then it seems you didn't include this script block.

Ditto.

As @AlexandreBulete says, just add it in yourself. You should see a hidden <input name="recaptcha_response" /> (or similar) in each form you want to use recaptcha in. Less convenient, but it sorts it.

**Comment by frostitution** [comment]
> Each form "Recaptchable" has an additional script block added. You have to watch the corresponding recaptcha module. For example : module-re-captcha-newsletter add this block :
> 
> ```
> <referenceContainer name="footer">
>             <block class="Magento\ReCaptchaUi\Block\ReCaptcha"
>                    name="recaptcha-newsletter"
>                    after="-"
>                    template="Magento_ReCaptchaNewsletter::recaptcha_newsletter.phtml"
>                    ifconfig="recaptcha_frontend/type_for/newsletter">
>                 <arguments>
>                     <argument name="recaptcha_for" xsi:type="string">newsletter</argument>
>                     <argument name="jsLayout" xsi:type="array">
>                         <item name="components" xsi:type="array">
>                             <item name="recaptcha" xsi:type="array">
>                                 <item name="component" xsi:type="string">Magento_ReCaptchaFrontendUi/js/reCaptcha</item>
>                             </item>
>                         </item>
>                     </argument>
>                 </arguments>
>             </block>
>         </referenceContainer>
> ```
> 
> Are you sure your layout is including this additional script block ?
> 
> I noticed it after dd() the `$request->getParams()` in this `CaptchaResponseResolver`model. If you haven't the requested recaptcha params then it seems you didn't include this script block.

I see this template on my site and I am still receiving the error for frontend Customer Registration

**Comment by rejith7** [comment]
@frostitution I'm getting the same error on the customer registration form(2.4.5-p4 EE). Were you able to resolve this?

**Comment by buderus199** [comment]
Same error for me on 2.4.5-p5 OS. Error occurs when bots try to register new customer account.

**Comment by krnshah** [comment]
I am having the same issue. We are using magento2.4.6-p2 + hyva. but when create the captcha key and apply it to frontend it show same error and not allowed us to login/sign up successfully.
Can anyone please help?

**Comment by frostitution** [comment]
@rejith7 No, sorry. I left it enabled because recaptcha works, but the log errors are still there. recaptcha works for us when we test it and cannot seem to replicate this consistently. I have been monitoring for any sort of solution from anyone but no luck.

**Comment by jorgb90** [comment]
Issue also still here!

**Comment by arvind-curotec** [comment]
Having same issue, no resolution.

**Comment by devvolunteeruk** [comment]
Having the same issue on Magento 2.4.7-p2, been seeing it since 2.4.6 on our installation

**Comment by engcom-Hotel** [comment]
Hello @devvolunteeruk @arvind-curotec,

We have tried to reproduce the issue in the latest development branch i.e. 2.4-develop but the issue is not reproducible for us. We have used Recaptcha v2 (I am not a robot).

There are no log entries with the specified error in the main description. Please let us know if we missed anything.

Thanks

**Comment by jmonrove** [comment]
Same issue with V3 on 2.4.6-p8

**Comment by engcom-Hotel** [comment]
Hello @jmonrove,

Please try to reproduce the issue either in the latest development branch i.e. [2.4-develop](https://github.com/magento/magento2.git) or the latest [stable version](https://experienceleague.adobe.com/en/docs/commerce-operations/release/versions).

We have tried to reproduce the issue in the development branch and it is not reproducible for us.

Thanks

**Comment by mcspronko** [comment]
The issue exists on Magento 2.4.6-p8 Open Source.

**Comment by engcom-Hotel** [comment]
Hello @mcspronko,

We recommend you please upgrade your instance to the latest stable version and let us know if you are still able to reproduce the issue.

Thanks

**Comment by mcspronko** [comment]
Hello @engcom-Hotel 

We are on the latest stable version of Magento. The 2.4.7-p3 has a bug with Cart Price Rules that we can't move to this version as of now.

**Comment by engcom-Hotel** [comment]
Thanks @mcspronko for the reply!

Are you facing the issue mentioned in the main description in 2.4.7-p3? If yes, can you please let us know the steps to reproduce it? Also, please try to reproduce the issue in the latest development branch, 2.4-develop and let us know if the issue is reproducible for you or not because for us the issue is not reproducible.

Thanks

**Comment by mcspronko** [comment]
@engcom-Hotel 

The 2.4.7-p3 has a bug with Cart Price Rules, not related to Captcha.

**Comment by engcom-Hotel** [comment]
Hello @mcspronko,

This issue has been created for the recaptcha bug and is no longer reproducible. Please create a new issue and explain the bug with the Cart price rules there. 

Thanks

**Comment by mcspronko** [comment]
@engcom-Hotel you asked me to update to the latest stable version where i replied that 2.4.7-p3 is not stable due to a bug with price cart rules. I am fully aware that the current issue is about captcha.

**Comment by mostafa-elkholy** [comment]
I have the same issue with Magento 2.4.6-p3 updated to 2.4.6-p8 & 2.4.7-p3 and have the same issue of the 

[2024-11-14T22:29:28.584459+00:00] main.ERROR: Magento\Framework\Exception\InputException: Can not resolve reCAPTCHA parameter. in /home/1218041.cloudwaysapps.com/rrkkmbpnmd/public_html/vendor/magento/module-re-captcha-ui/Model/CaptchaResponseResolver.php:25
Stack trace:
#0 vendor/magento/module-re-captcha-ui/Model/RequestHandler.php(112): Magento\ReCaptchaUi\Model\CaptchaResponseResolver->resolve()
#1 vendor/magento/module-re-captcha-customer/Observer/CreateCustomerObserver.php(67): Magento\ReCaptchaUi\Model\RequestHandler->execute()
#2 vendor/magento/framework/Event/Invoker/InvokerDefault.php(88): Magento\ReCaptchaCustomer\Observer\CreateCustomerObserver->execute()
#3 vendor/magento/framework/Event/Invoker/InvokerDefault.php(74): Magento\Framework\Event\Invoker\InvokerDefault->_callObserverMethod()
#4 vendor/magento/framework/Event/Manager.php(65): Magento\Framework\Event\Invoker\InvokerDefault->dispatch()
#5 generated/code/Magento/Framework/Event/Manager/Proxy.php(95): Magento\Framework\Event\Manager->dispatch()
#6 vendor/magento/framework/App/FrontController.php(279): Magento\Framework\Event\Manager\Proxy->dispatch()
#7 vendor/magento/framework/App/FrontController.php(211): Magento\Framework\App\FrontController->dispatchPreDispatchEvents()
#8 vendor/magento/framework/App/FrontController.php(147): Magento\Framework\App\FrontController->processRequest()
#9 vendor/magento/framework/Interception/Interceptor.php(58): Magento\Framework\App\FrontController->dispatch()
#10 vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\App\FrontController\Interceptor->___callParent()
#11 vendor/nitropack/magento2-extension/Plugin/CacheDelivery/LocalCachePlugin.php(183): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}()
#12 vendor/magento/framework/Interception/Interceptor.php(135): NitroPack\NitroPack\Plugin\CacheDelivery\LocalCachePlugin->aroundDispatch()
#13 vendor/magento/module-store/App/FrontController/Plugin/RequestPreprocessor.php(99): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}()
#14 vendor/magento/framework/Interception/Interceptor.php(135): Magento\Store\App\FrontController\Plugin\RequestPreprocessor->aroundDispatch()
#15 vendor/magento/module-page-cache/Model/App/FrontController/BuiltinPlugin.php(71): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}()
#16 vendor/magento/framework/Interception/Interceptor.php(135): Magento\PageCache\Model\App\FrontController\BuiltinPlugin->aroundDispatch()
#17 vendor/magento/framework/Interception/Interceptor.php(153): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}()
#18 generated/code/Magento/Framework/App/FrontController/Interceptor.php(23): Magento\Framework\App\FrontController\Interceptor->___callPlugins()
#19 vendor/magento/framework/App/Http.php(116): Magento\Framework\App\FrontController\Interceptor->dispatch()
#20 generated/code/Magento/Framework/App/Http/Interceptor.php(23): Magento\Framework\App\Http->launch()
#21 vendor/magento/framework/App/Bootstrap.php(264): Magento\Framework\App\Http\Interceptor->launch()
#22 pub/index.php(30): Magento\Framework\App\Bootstrap->run()

**Comment by bjolysharp** [comment]
Why is this issue closed? We're still seeing the issue on Magento 2.4.7-p3:

`[2025-01-03T13:21:50.059509+00:00] main.ERROR: Magento\Framework\Exception\InputException: Can not resolve reCAPTCHA parameter. in /chroot/home/.../html/vendor/magento/module-re-captcha-ui/Model/CaptchaResponseResolver.php:25`

**Comment by alexandresoaresramos** [comment]
2.4.7 p3

{main} {"exception":"[object] (Magento\\Framework\\Exception\\InputException(code: 0): Can not resolve reCAPTCHA parameter. at /home/viprac/public_html/vendor/magento/module-re-captcha-ui/Model/CaptchaResponseResolver.php:25)"} []

**Comment by drabikowy** [comment]
We have the same issue, I don't know why this ticket was closed with so many reports about issue existence. 
Logs are spammed for us with the same errors.

**Comment by hardikkukadiya** [comment]
Give me solution for stop generating error log

**Comment by PieterCappelle** [comment]
Any update about this or a solution?

**Comment by gntlby** [comment]
main.ERROR: Magento\Framework\Exception\InputException: Can not resolve reCAPTCHA parameter. in .../vendor/magento/module-re-captcha-ui/Model/CaptchaResponseResolver.php:25

We are experiencing the same issue on Magento 2.4.7-p4. 

We would appreciate support regarding the solution.

**Comment by hostep** [comment]
Would [patch MDVA-39305(or -V2 or -V3)](https://github.com/magento/quality-patches/blob/73ae870025d694cf39e2de6ffa3f63f24ca9efc8/support-patches.json#L3722-L3861) help out here?
One of the errors they associate with that patch is the same as the one reported here. See: https://github.com/magento/quality-patches/blob/73ae870025d694cf39e2de6ffa3f63f24ca9efc8/tests/metadata/MDVA-39305-V3/var/log/support_report.log

Just guessing here ...

**Comment by devchris79** [comment]
@hostep Thanks for that, we have seen this error and it was definitely correlating with a customer struggling to login, oddly switching browser sorted the issue at the time.
I have applied the quality patch and will see if the error is resolved.

**Comment by devchris79** [comment]
@frostitution we had a couple more stamps on the evening after applying the patch, they were 28 seconds apart, so I would assume it was a result of a login attempt failure.
Our site has a low traffic level, so its hard to get a feel for if there has been a reduction in errors or no effect at all.

It may be a case that there is multiple ways that this error is being triggered.

**Comment by devchris79** [comment]
There has been no further errors after appying patch MDVA-39305-V3 as suggested above by @hostep. I am wondering if the errors that I saw on the same day as applying the patch were the result of an old cache data.
I will post if we get any more errors, but I think the possible fix needs checking by others. 
Given the fact the patch fixes login errors, it seems a no brainer to install.

**Comment by devchris79** [comment]
Unfortunately the stamp has occured in the exception log again this morning:

[2025-04-01T06:23:45.675148+00:00] main.ERROR: Magento\Framework\Exception\InputException: Can not resolve reCAPTCHA parameter. in ../vendor/magento/module-re-captcha-ui/Model/CaptchaResponseResolver.php:25

**Comment by bolasevich** [comment]
i also have this issue

**Comment by hewgim** [comment]
Hi @devchris79, Are you seeing issues on the frontend because this ticket is specifically about trouble with google recaptcha when logging into the backend, which the @tschirmer code seems to have fixed and as far as I can see. And this code has been incorporated into the 2.4.6 branch as far as I can see.

Are people are still having issues with Google Recacptcha when logging into the backend ? If not then this ticket should probably be closed.

Regards

Mark

**Comment by devchris79** [comment]
@hewgim The @tschirmer code seemed to fix the admin panel login errors for us. I can't speak for the main branch code fix as its not something we have tested.

If this ticket is to be closed a new one opened for the frontend issues, a link for users with the frontend ReCaptcha errors would be appreciated.

**Comment by bolasevich** [comment]
yes im still having frontend issues as well

**Comment by engcom-Bravo** [comment]
Hi @stevekem,

Internal Team investigated this issue in our development environment and tested the scenario across the following Magento versions:

2.4.8-develop
2.4.7-p3
2.4.6-p8

These versions were specifically mentioned by community users in the related GitHub thread. Unfortunately, we were unable to replicate the exact behavior described in the ticket under normal conditions where the user is genuine.
Observations:

When a bot or a user incorrectly completes the reCAPTCHA, the system throws an InputException and logs the entire object.
This behavior is due to how the code is currently written, and it results in excessive entries in the exception.log file.
Some users reported login failures on the first or second attempt, but success on the third. We were not able to replicate this behavior.

The issue appears to be tied to failed reCAPTCHA validations, likely triggered by bots accessing login or forgot password pages.
Request for Reproduction Steps:

If any community member can attach a video demonstrating the issue step by step, that would be extremely helpful. The current ticket lacks details about the specific actions taken by the user. A clear reproduction path would allow us to investigate more effectively and validate the issue under real-world conditions.

Thanks.

**Comment by regressEdo** [comment]
we are also facing the same issue, starting from 21st of June, we are daily spammed of `Magento\\Framework\\Exception\\InputException: Can not resolve reCAPTCHA parameter. in /srv/vendor/magento/module-re-captcha-ui/Model/CaptchaResponseResolver.php:25`

current magento version: 2.4.6-p8

**Comment by devchris79** [comment]
@engcom-Bravo There is an easy to implement example of how to trigger the log stamps provided by @barryvdh above.
Can you please retest and add the issue confirmed labelling (assuming you replicate the issue).

This has spent over 3 years as a P1 issue through multiple versions, so it would be great if the internal team could review.

**Comment by tschirmer** [comment]
Essentially, the issue that I'd provided a solve for here: https://github.com/magento/magento2/issues/34975#issuecomment-1138038548 - that got added to the codebase, was because the order of operations for conducting the recaptcha weren't respected. 

The same issue exists for the frontend code, I've haven't spent the time to fix it yet. When I reviewed it, it appeared to be the same issue.

**Comment by barryvdh** [comment]
I don't think there really is an issue, except that bots are trying to login. That is correctly prevented with the captcha, but now it fills the exception log.

**Comment by hostep** [comment]
@engcom-Bravo: the security package is publicly available, I think it's fixed with https://github.com/magento/security-package/commit/804dbc2ad49d6adc73d93795ca7ccfe5e6d5f775 (ticket name: "AC-3179: "Can not resolve reCAPTCHA parameter" entries in exception.log for Google reCAPTCHA Admin Panel"). Can you confirm this?

**Comment by engcom-Bravo** [comment]
Hi @hostep,

Thanks for your contribution!!.

yes you are correct it is fixed under this commit https://github.com/magento/security-package/commit/804dbc2ad49d6adc73d93795ca7ccfe5e6d5f775

Thanks.

**Comment by JMLucas96** [comment]
> [@engcom-Bravo](https://github.com/engcom-Bravo): the security package is publicly available, I think it's fixed with [magento/security-package@804dbc2](https://github.com/magento/security-package/commit/804dbc2ad49d6adc73d93795ca7ccfe5e6d5f775) (ticket name: "AC-3179: "Can not resolve reCAPTCHA parameter" entries in exception.log for Google reCAPTCHA Admin Panel"). Can you confirm this?

Same issue here with Magento 2.4.8-p5 solved with that MR. The problem was that captcha throws error if navigation is not legitimate like fraud bots, DDOS attacks, etc...
